src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -30,6 +34,192 @@
+      for fn in provider.resource_listdir(package):
+        if fn.endswith('.py'):
+          rel_path = os.path.join(package, fn)
+          self._chroot.write(provider.get_resource_string(source_name, rel_path),
+                             os.path.join(BOOTSTRAP_DIR, source_name, rel_path),
+                             'bootstrap')
+
+    # Setup a re-director package to support the legacy pex runtime `_pex` APIs through a
+    # VendorImporter.
+    self._chroot.touch(os.path.join(BOOTSTRAP_DIR, LEGACY_BOOSTRAP_PKG, '__init__.py'), 'bootstrap')
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+if ext == '.pex':
+  raise ValueError("Could not hydrate ipex: {}.pex must not be named with the .pex extension.".format(filename_base))
+hydrated_pex_file = '{}.pex'.format(filename_base)
+
+if not os.path.isfile(hydrated_pex_file):
+  print('Hydrating {} to {}'.format(self, hydrated_pex_file))
+
+  ipex_info = PexInfo.from_pex(self)
+  if ipex_info.zip_safe:
+    raise ValueError("Could not hydrate ipex: ipex info must be marked zip_safe=False. This ipex file's info was:\\n{}".format(ipex_info.dump(indent=4)))
+  code_root = os.path.join(ipex_info.zip_unsafe_cache, ipex_info.code_hash)
+  with open_zip(self) as zf:
+    # Populate the pex with the pinned requirements and distribution names & hashes.
+    orig_pex_info = PexInfo.from_json(zf.read('IPEX-INFO'))
+    ipex_builder = IPEXBuilder(pex_info=orig_pex_info)
+
+    # Populate the pex with the needed code.
+    ptex_info = json.loads(zf.read('PTEX-INFO').decode('utf-8'))
+    for path in ptex_info['code']:
+      ipex_builder.add_source(os.path.join(code_root, path), path)
+
+    # Get the requirements to resolve.
+    plat = Platform.current()
+    transitive_requirements_by_platform = ptex_info['transitive_requirements_by_platform']
+    if plat.platform not in transitive_requirements_by_platform:
+      raise ValueError("Current platform {} was not found in the 'transitive_requirements' field in PTEX-INFO:\\n{}".format(plat, json.dumps(ptex_info, indent=4)))
+
+    # Hack the vendor specs so that they point to existing directories so that they can be copied to
+    # the IPEXBuilder chroot.
+    setuptools_spec = vendor.VendorSpec.create('setuptools==40.6.2')
+    wheel_spec = vendor.VendorSpec.create('wheel==0.31.1')
+    vendored_specs = [setuptools_spec, wheel_spec]
+
+    for spec in vendored_specs:
+      if os.path.isdir(spec.target_dir):
+        raise ValueError("Vendored specs are expected to be within the zipped pex, but {} from {} is a real directory!".format(spec.target_dir, spec))
+      files_in_dir = [name for name in zf.namelist()
+                      if name.startswith(os.path.join(BOOTSTRAP_DIR, spec.relpath))]
+      zf.extractall(code_root, files_in_dir)
+      spec.ROOT = os.path.join(code_root, BOOTSTRAP_DIR)
+    # NB: Monkey-patch the pex.vendor.iter_vendor_specs() method!
+    vendor.iter_vendor_specs = lambda: vendored_specs
```

Comment:
From ~here up can be deleted if you just include a `pex` requirement in the ptex as was done in the sample here: https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R192

Reply:
Thank you so much! I had been having trouble parsing that command line in my head!